library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java

```diff
@@ -442,9 +443,16 @@ private Slice readOrSkip(boolean skip) throws IOException
           inputStream = getInputStream(currentPointer);
         }
 
-        if (inputStream != null && currentPointer.offset <
-            fileSystemWAL.fileContext.getFileStatus(currentOpenPath).getLen()) {
-          int len = inputStream.readInt();
+        int ch1;
+
+        if (inputStream != null && (ch1 = inputStream.read()) != -1) {
```

Comment:
The reference here shouldn't be the book but the existing API that we depend on and need to work with optimally. So if relying on the EOFException raised would be more suitable, then we should do so. @vrozov any thoughts?

Overall this fix is critical and we agree on the approach now, so we don't need to delay over splitting hairs.